src/contract.rs

```diff
@@ -0,0 +1,468 @@
+            )
+            .await?;
+            return Err(error);
+        }
+
+        // Painting
+        paint(
+            self.resolution,
+            &rectangle,
+            rectangle_width,
+            rectangle_height,
+            &mut self.painting,
+            painting,
+        );
+        let token_info = self
+            .tokens_by_rectangles
+            .entry(rectangle)
+            .or_insert_with(|| TokenInfo {
+                owner: msg_source,
+                pixel_price: None,
+                token_id: None,
+            });
+
+        let token_id = utils::mint_nft(tx_id, &self.nft_program, token_metadata).await?;
+        tx_id = tx_id.wrapping_add(1);
+        utils::transfer_nft(tx_id, &self.nft_program, &msg_source, token_id).await?;
+        utils::transfer_ftokens(
+            tx_id,
+            &self.ft_program,
+            &exec::program_id(),
+            &self.owner,
+            rectangle_pixel_count as u128 * self.pixel_price,
+        )
+        .await?;
+        // Insertion and replying
+        token_info.token_id = Some(token_id);
+        self.rectangles_by_token_ids.insert(token_id, rectangle);
+        Ok(NFTPixelboardEvent::Minted(token_id))
+    }
+
+    async fn buy(
+        &mut self,
+        mut tx_id: TransactionId,
+        token_id: TokenId,
+    ) -> Result<NFTPixelboardEvent, NFTPixelboardError> {
+        let msg_source = msg::source();
+        let (rectangle, token) = get_mut_token(
+            &self.rectangles_by_token_ids,
+            &mut self.tokens_by_rectangles,
+            token_id,
+        )?;
+
+        let pixel_price = if let Some(pixel_price) = token.pixel_price {
+            pixel_price
+        } else {
+            return Err(NFTPixelboardError::NFTIsNotOnSale);
+        };
+
+        // get_pixel_count() isn't used here because it checks an NFT area for
+        // equality to 0, but here it's always not equal 0.
+        let token_price =
+            (rectangle.width() as usize * rectangle.height() as usize) as u128 * pixel_price;
+        let resale_commission = token_price * self.commission_percentage as u128 / 100;
+
+        utils::transfer_ftokens(
+            tx_id,
+            &self.ft_program,
+            &msg_source,
+            &exec::program_id(),
+            token_price,
+        )
+        .await?;
+
+        tx_id = tx_id.wrapping_add(1);
+
+        utils::transfer_ftokens(
+            tx_id,
+            &self.ft_program,
+            &exec::program_id(),
+            &self.owner,
+            resale_commission,
+        )
+        .await?;
+
+        tx_id = tx_id.wrapping_add(1);
+
+        utils::transfer_ftokens(
+            tx_id,
+            &self.ft_program,
+            &exec::program_id(),
+            &token.owner,
+            token_price - resale_commission,
+        )
+        .await?;
+
+        utils::transfer_nft(0, &self.nft_program, &msg_source, token_id).await?;
```

Comment:
Why is `tx_id` not passed as transaction_id, but `0` instead?

Reply:
My mistake